### PR TITLE
Replace std::array with Kokkos::Array in Tensor

### PR DESCRIPTION
**Pull request by Rombur**
I cannot compile a bunch of tests in debug mode with HIP. Everything works fine in release mode but in debug, I get errors because we use `std::array` on the device:
```
/usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/array:202:2: error: reference to __host__ function '__glibcxx_assert_fail' in __host__ __device__ function 
  202 |         __glibcxx_requires_subscript(__n);           
      |         ^      
/usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/debug/assertions.h:44:3: note: expanded from macro '__glibcxx_requires_subscript'               
   44 |   __glibcxx_assert(_N < this->size())           
      |   ^        
/usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/x86_64-linux-gnu/c++/13/bits/c++config.h:612:8: note: expanded from macro '__glibcxx_assert'                        
  612 |   do { __glibcxx_assert_impl(cond); } while (false)     
      |        ^       
/usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/x86_64-linux-gnu/c++/13/bits/c++config.h:597:10: note: expanded from macro '__glibcxx_assert_impl'      
  597 |     std::__glibcxx_assert_fail(__FILE__, __LINE__, __PRETTY_FUNCTION__, \    
      |          ^   
/home/dealii/include/deal.II/base/tensor.h:1457:10: note: called by 'operator[]'       
 1457 |   return values[i];
```
Replacing `std::array` with `Kokkos::Array` fixes the issues. I had to change the serialization function because we were using boost. The new code works but the serialization is different.

### CODE REVIEW

File: include/deal.II/base/tensor.h [hunk 1]
```diff
@@ -24,21 +24,12 @@
 #include <deal.II/base/template_constraints.h>
 #include <deal.II/base/tensor_accessors.h>
 
+#include <Kokkos_Core.hpp>
+
```
Comment (drwells) [inline include/deal.II/base/tensor.h:28]:
```suggestion
#include <Kokkos_Array.hpp>

```

Can we get away with using that instead?

File: include/deal.II/base/tensor.h [hunk 2]
```diff
@@ -1262,17 +1253,17 @@ template <int rank_, int dim, typename Number>
 constexpr DEAL_II_HOST_DEVICE_ALWAYS_INLINE
 Tensor<rank_, dim, Number>::Tensor()
   : values(
-      // In order to initialize the std::array<Number,dim>, we would need a
+      // In order to initialize the Kokkos::Array<Number,dim>, we would need a
       // brace-enclosed list of length 'dim'. There is no way in C++ to create
       // such a list in-place, but we can come up with a lambda function that
       // expands such a list via template-pack expansion, and then uses this
-      // list to initialize a std::array which it then returns.
+      // list to initialize a Kokkos:Array which it then returns.
```
Comment (drwells) [inline include/deal.II/base/tensor.h:1260]:
```suggestion
      // list to initialize a Kokkos::Array which it then returns.
```

**Review by drwells — COMMENTED** [review]
Interesting problem - it looks like the assertions (which we turned on with `_GLIBCXX_DEBUG_ASSERTIONS`) don't work with GPUs.

**Comment by kronbichler** [comment]
Does your error message mean that we would need to disable GLIBCXX assertions like we do for CUDA? We have https://github.com/dealii/dealii/blob/4f432e494ab19ea38078379f6644f142aad4ac5d/cmake/modules/FindDEAL_II_KOKKOS.cmake#L63-L65 and your error messages sound very similar to what I'd expect there. But I am also not opposed to the present change.

**Comment by Rombur** [comment]
> Does your error message mean that we would need to disable GLIBCXX assertions for HIP like we do for CUDA?

I didn't realize we were disabling GLIBCXX assertions in CUDA. It would probably work too but I would prefer not having to disable the assertions. Let me see if I can get all the compilers happy.

**Comment by Rombur** [comment]
It's working now. We can only use `Kokkos::Array` if Kokkos is new enough (>= 3.7). Before that version, `operator[]` was not `constexpr`. I also had to slightly modify the initialization of Tensor because of a small difference in list initialization between `std::array<T, 0>` and `Kokkos::Array<T, 0>` (`std::array<T,0> a = {{}}` is legal whereas for `Kokkos::Array`, we need to use `Kokkos::Array<T, 0> a = {}`)

**Review by masterleinad — APPROVED** [review]
Fine with me.

**Review by kronbichler — COMMENTED** [review]
I also think this should be good to go. (Assuming that there are no implicit additional data members in the array, which I did not see when looking over `Kokkos::array`).

**Comment by Rombur** [comment]
> Assuming that there are no implicit additional data members in the array

No it's just a c-array with the member functions marked with `__host__ __device__`

File: cmake/modules/FindDEAL_II_KOKKOS.cmake [hunk 3]
```diff
@@ -47,6 +47,15 @@ endif()
...
 if(Kokkos_FOUND)
+  if (Kokkos_ENABLE_CUDA OR Kokkos_ENABLE_HIP)
+    # In version older than 3.7.0, Kokkos::Array::operator[] is not constexpr,
+    # so we use std::array instead.
+    if (KOKKOS_VERSION VERSION_LESS 3.7.0)
+      # We are using std::array in device code which calls the host-only function
+      # __glibcxx_requires_subscript when defining _GLIBCXX_ASSERTIONS 
+      list(REMOVE_ITEM DEAL_II_DEFINITIONS_DEBUG "_GLIBCXX_ASSERTIONS")
+    endif()
+  endif()
```
Comment (bangerth) [inline cmake/modules/FindDEAL_II_KOKKOS.cmake:58]: Is this still necessary now?

File: cmake/modules/FindDEAL_II_KOKKOS.cmake [hunk 4]
```diff
@@ -60,10 +69,6 @@ if(Kokkos_FOUND)
     # __host__ constexpr functions in device code
     KOKKOS_CHECK(OPTIONS CUDA_LAMBDA CUDA_CONSTEXPR)
 
-    # We are using std::array in device code which calls the host-only function
-    # __glibcxx_requires_subscript when defining _GLIBCXX_ASSERTIONS 
-    list(REMOVE_ITEM DEAL_II_DEFINITIONS_DEBUG "_GLIBCXX_ASSERTIONS")
-
```
Comment (bangerth) [inline cmake/modules/FindDEAL_II_KOKKOS.cmake:66]: Same here?

Comment (Rombur) [reply cmake/modules/FindDEAL_II_KOKKOS.cmake:66]: These lines are removed.

File: cmake/modules/FindDEAL_II_KOKKOS.cmake (hunk 3 above)
Comment (bangerth) [reply cmake/modules/FindDEAL_II_KOKKOS.cmake:58]: Oh, I see now. Yes, that makes sense.

**Pull request merged**